### ISSUE TITLE
Save and restore questions for RaftStore and StateMachine

### ISSUE THREAD

**Issue by momiji** [MicroRaft/MicroRaft#18]
Hello,

I'm quite interested in implementing microraft in one of my internal project, however I'm not sure I did understand everything, even after several read and re-read, and 2 nights thinking about it :)

If you're not too busy to help, it will me much appreciated!

To better understand the library, I started to replay part of the tutorial, oriented on my specific needs (let's say a sort of kv to keep it simple):
- run in multiple jvm instead of multi-thread, to have serialization and network
- network is very basic socket to keep it simple until I have everything I want working
- have high availability and failover and resiliency...

My questions at this time is around the RaftStore and StateMachine.
I understand I need to implement recovery in RaftStore and provide RestoredRaftState on startup, but not sure what to save (on disk at first) to implement this.
It seems one need to save (so we can restore): localEndpoint, localEnpointVoting, initialGroupMembers and term. All these are easy to do.
Then comes the log entry and snapshot chunks, and here I'm quite lost, because from the only implemtation I can see (InMemoryRaftStore), it is using a RaftLog to do so, but the code looks weird to me:
- persistSnapshotChunk: is the code reusable like it or should I so something completely different?
- truncateSnapshotChunksUntil: I'm not sure clear() is the good thing I should do...
I had the impression this was a fast go to implement the store, but I may be wrong.

Concerning the StateMachine, I wonder wether I should also save the data to disk, or would everything be replayed?
Does it mean that during Store save, I should also save the state machine, or is ti 2 totally different things to do (as per the doc), and if so when should it happen?

```java
    @Override
    public void persistSnapshotChunk(SnapshotChunk snapshotChunk) throws IOException {
        snapshotChunks.add(snapshotChunk);

        if (snapshotChunk.getSnapshotChunkCount() == snapshotChunks.size()) {
            snapshotChunks.sort(comparingInt(SnapshotChunk::getSnapshotChunkIndex));
            SnapshotEntry snapshotEntry = new DefaultSnapshotEntryOrBuilder().setTerm(snapshotChunk.getTerm())
                    .setIndex(snapshotChunk.getIndex())
                    .setSnapshotChunks(snapshotChunks)
                    .setGroupMembersView(
                            snapshotChunk.getGroupMembersView())
                    .build();
            raftLog.setSnapshot(snapshotEntry);
            snapshotChunks = new ArrayList<>();
        }
    }

    @Override
    public void truncateSnapshotChunksUntil(long logIndexInclusive) throws IOException {
        snapshotChunks.clear();
    }
```

Regards,

**Comment by metanet** [comment]
Hi @momiji 

Thanks for your interest in MicroRaft. 

I wrote InMemoryRaftStore just to build a simple demo, but normally a RaftStore implementation must persist local endpoint information, term information, log entries and snapshots to persistent storage by following the API. Please ignore InMemoryRaftStore if it is causing confusion. 

StateMachine can be completely reside in memory. On restart / recovery, it will be populated with the information provided in RestoredRaftState. Basically, MicroRaft will read the latest snapshot and log entries from disk and will replay them to build the same state in the StateMachine implementation.

Your RaftStore implementation must persist local endpoint, term information, log entries and snapshots to the disk.  For persisting log entries and snapshots, a flush() method is separately provided so that for instance MicroRaft will first call persistLogEntry() for N log entries and then it will call flush() and under the hood you can do batching / buffering etc. to get better performance. But once flush() is called, your implementation must guarantee that previous persist...() or truncate...() calls are now durably persisted. 

MicroRaft divides snapshots into chunks. For simplicity, you can return only a single snapshot chunk from your StateMachine.takeSnapshot() implementation. When such a chunk object is sent by leader to a follower, the follower RaftNode will call RaftStore.persistSnapshotChunk(). Once the follower receives all chunks of the latest snapshot, it will call RaftStore.flush() and at this point your implementation must guarantee durability of those chunks, i.e., it must be able to restore them in case of a crash. But suppose that there are N chunks from the latest snapshot. A follower persisted some of these N chunks but in the meantime the leader took a new snapshot. In this case, on the next request the follower will receive a snapshot chunk with a higher index, and then it will call RaftStore.truncateSnapshotChunksUntil() to denote that all non-flushed chunk objects up to the passed index are no longer needed.

For instance, you can persist each snapshot to a separate file. For e.g., snapshot taken at log index=N can be persisted to `snapshot-N.raft` etc. When you receive a snapshot chunk, you can first create it as ``snapshot-N.raft.tmp`. After all chunks are persisted to this file, on flush() call you can atomically rename it to `snapshot-N.raft`. If truncate() method is called, you can just delete the temp file. 

I hope my reply helps.

**Comment by momiji** [comment]
Thanks a lot for you fast reply, I'll give it a try this way and see where it leads me.

> MicroRaft divides snapshots into chunks. For simplicity, you can return only a single snapshot chunk from your StateMachine.takeSnapshot() implementation.

While reading your reply, I finally understood that I completely failed to see what snapshot are, but hopefully it is now more clear to me.
And it explains the persistSnapshotChunk() code in the InMemoryRaftStore (espcially the size test), as well as the truncateSnapshotChunksUntil() which juste clear the list, as if it was a new file.

Another quick question, are you aware of projects already using this library so far?

Anyway, it was a very intersting study of what raft is, and if I can achieve integration like I want, I will probably get rid of consul and our internal project.

Regards,

**Comment by metanet** [comment]
there were a few folks playing with the library but I am not aware of any prod deployment yet. I am also building a simple kv store impl but it is taking more time than I thought :(

**Comment by momiji** [comment]
I found my first headache on the implementation with files backend, but I think I finllaly got it!

I splitted the RaftStore storage in two: NodeInfo for everything related to node, having persistSomething() method, and a StoreInfo to save the latest snapshot and the log entries.

From my experience, it is required on restore to fill manually these objects (well, at least the snapshot and log entries) otherwise everything is broken after 2 restarts.

Original non working code:

```java
    public synchronized RestoredRaftState toRestoredRaftState(String nodeFilename, String storeFilename) throws IOException {
        // read files from disk
        if (!Files.exists(Paths.get(nodeFilename))) return null;
        NodeInfo node = NodeInfo.load(nodeFilename);
        StoreInfo store = null;
        if (Files.exists(Paths.get(storeFilename))) {
            store = StoreInfo.load(storeFilename);
        }
        // compute snapshot and log entries
        SnapshotEntry snapshotEntry = store != null ? store.snapshotEntry : null;
        List<LogEntry> logEntries = store != null ? store.logEntries : new ArrayList<>();
        // return state for library internals
        return new RestoredRaftState(node.localEndpoint, node.localEndpointVoting, node.initialGroupMembers, node.term,
                node.votedFor, snapshotEntry, logEntries);
    }
```

Final working code, with the following modifications:
- filter log entries to keep only logs with index >= snapshotIndex
- save the snapshot and log entries into my RaftStore implementation

```java
    public synchronized RestoredRaftState toRestoredRaftState(String nodeFilename, String storeFilename) throws IOException {
        // read files from disk
        if (!Files.exists(Paths.get(nodeFilename))) return null;
        NodeInfo node = NodeInfo.load(nodeFilename);
        StoreInfo store = null;
        if (Files.exists(Paths.get(storeFilename))) {
            store = StoreInfo.load(storeFilename);
        }
        // compute snapshot and log entries filtered to keep only log entries after snapshot
        SnapshotEntry snapshotEntry = store != null ? store.snapshotEntry : null;
        List<LogEntry> logEntries = store != null ? store.logEntries : new ArrayList<>();
        long snapshotIndex = snapshotEntry != null ? snapshotEntry.getIndex() : 0;
        logEntries = logEntries.stream().filter(e -> e.getIndex() >= snapshotIndex).collect(Collectors.toList());
        // update store with snapshot and log entries 
        storeInfo.snapshotEntry = snapshotEntry;
        storeInfo.logEntries = logEntries;
        // return state for library internals
        LOGGER.info("restored entries : {}", logEntries.size());
        LOGGER.info("restored snapshot : {}", snapshotEntry);
        return new RestoredRaftState(node.localEndpoint, node.localEndpointVoting, node.initialGroupMembers, node.term,
                node.votedFor, snapshotEntry, logEntries);
    }
```

Maybe this can be added to a sort of FAQ or an issue called "FAQ: restoring snapshots and log entries"

**Comment by momiji** [comment]
I have another question for the flush() in Store: it can see it is happening each time a new log entry is added.

I'm looking for a way to optimize this and reduce the amount of disk writes per second.

My need is something like this:
- for certain operation, we need to save immediately to disk (i.e. we create a new object in the kv)
- for other, we can afford to lose data so we can save them later (i.e. counters, eta, ...)

Is it something as simple as altering the flush() to deal with that:
- foreach log entry to be saved, if it contains a snapshot entry or a critical operation, save immediately
- otherwise, postpone the save action

The only remaining action will be to fire a save every 10s, but I guess this can be simply handled by adding a noop operation to do that, allowing to not care about thread safety and so on :)

Do you see any issue with that?

**Comment by metanet** [comment]
@momiji 

I am pretty certain that if you alter the logic of flush() call you will break the correctness of the algorithm. flush() must provide the requirement specified in the RaftState javadocs. It is due to Raft itself. When a follower replies to a leader for a log entry, from that moment it must have the information about that log entry for sure, hence it must flush() before its reply. 

To amortize the cost of flush() calls MicroRaft has batching, pipelining and parallel flush() calls:
- Leader sends log entries to followers in batches,
- While followers process append requests, leader can append new log entries,
- For a single log entry, leader and followers call flush() in parallel, not serially.

Please check [the config knobs](https://github.com/MicroRaft/MicroRaft/blob/1588e9d5cd37d6aa541792620f2331c8fb8e9fb8/microraft/src/main/java/io/microraft/RaftConfig.java#L114) and try different settings. You can monitor how many log entries are being sent in append requests. I was hoping for someone to integrate the micrometer library actually :)

I will check your other q later.

cheers,

**Comment by momiji** [comment]
Hello,

I was not clear on my explanation at all, really sorry about that.

**Flush and disk writing**
My idea was to delay write on disk for specific operations that can be lost (which I called non-critical), not to alter the way the flush is working (which will definetly break raft, for sure).
I've made a proof of concept for it and it worked very well and fits my need for this.

The global idea is:
- critical op => write log entries on disk
- non-critical op => start timer
- non-critical op => start timer (does nothing as timer already started)
- non-critical ops...
- critical op => write log entries on disk and stop timer
- non-critical op => start timer
- non-critical op => start timer (does nothing as timer already started)
- non-critical ops...
- on timer => add timer critical op to force writing log entries
- timer critical op => write log entries on disk
- ...

**Serialization**
I also had the same feeling at first as #8, as there is some code to write to make everything serializable, and more important to make everything deserializable without braking the app. Because any change in the code (microraft objects as well as my objects) having the effect of changing serialVersionUID in serializable classes which leads to the impossibility to load objects fron the log entries written on disk, I finally decided to write serialization/deserialization code for all serialiazed objects (readObject/writeObject) that will go to disk, but this is quite easy to do, so it was just a matter of some hours of work.

Nonetheless, even if the tests you've made are all done in a multithread environment, I was able to have everything working in a multiprocess environment without too much trouble, except my comment above https://github.com/MicroRaft/MicroRaft/issues/18#issuecomment-1109075368 about restoring snapshot, which was the hardest thing to find out. Everything else went smoothly.

**Builder wow!**
I was also definietly impressed by the way you wrote your classes, the builder, and so on, as I was able to rebuild all necessary objects using the builder very easily, don't know if you're using any code plugin to generate the builder, but this is wow!!! I wish I could do the same by my myself :)

**First tests**
I now have a running poc with encrypted grpc, and was able to achieve a good throughput (near 800 ops/s) on a single server with 3 separate processes, my next test will be on a stretched cluster, again with 3 nodes, and then with 5 nodes to see if there is a performance penalty (which I'm sure there will be) and of how much.

**Future tests**
I'm expecting to have a deployement with 3 or 5 servers:
- 2 servers on datacenter 1
- 2 servers on datacenter 2 - our app is on a stretched cluster (several kms), with near 1ms latency between the 2 DC (dedicated fiber link)
- 1 server somewhere else, more like a witness or something like that, just for ensuring we can lose one DC without loosing the app. Doing so should not have any impact, as long as pushing new log entries will return as soon as "qorums" servers have received the entries, but this is something I need to confirm and test. Am I right?

Thanks again for the time you're taking to answer my (not so well english-written) questions :)

**Comment by metanet** [comment]
> Serialization

https://github.com/MicroRaft/MicroRaft/tree/master/microraft/src/main/java/io/microraft/model this package mostly works with builders and interfaces to enable developers to use different serde and rpc protocols under the hood. For e.g., you can use protobuf to represent MicroRaft's model interfaces and wrap them in the implementations of the model interfaces. Then those wrappers can delegate the set and get calls to the underlying protobuf message objects. See the following example:

```
public class LogEntryOrBuilder implements LogEntry, LogEntryBuilder {

    private LogEntryProto.Builder builder;
    private LogEntryProto entry;

    public LogEntryOrBuilder() {
        this.builder = LogEntryProto.newBuilder();
    }

    public LogEntryProto getEntry() {
        return entry;
    }

    public LogEntryOrBuilder(LogEntryProto entry) {
        this.entry = entry;
    }

    public LogEntryProto getEntry() {
        return entry;
    }

    @Nonnull
    @Override
    public LogEntryBuilder setIndex(long index) {
        builder.setIndex(index);
        return this;
    }

    // other setter impls to populate this.builder's fields

    @Nonnull
    @Override
    public LogEntry build() {
        entry = builder.build();
        builder = null;
        return this;
    }

    @Override
    public String toString() {
        if (builder != null) {
            return "ProtobufLogEntry{builder=" + builder + "}";
        }

        return "ProtobufLogEntry{" + "index=" + getIndex() + ", term=" + getTerm() + ", operation=" + getOperation() + '}';
    }

    @Override
    public long getIndex() {
        return entry.getIndex();
    }

    @Override
    public int getTerm() {
        return entry.getTerm();
    }

    // other getter methods.
}
```
Then in your transport implementation you can extract protobufs objects from these wraps send send them over the wire.

> Future tests

MicroRaft doesn't support the witness feature now. I am not sure if it worths the complexity because I think it brings a bit of complexity to the quorum calculations. But if there is a real need we can check again. IF we assume witness functionality is available in MicroRaft, the scenario you describe is ok.

**Comment by momiji** [comment]
Yes, for serialization, this is exactly what I'm doing.

For each object, I just get the fields values needed by the builder, serialize them "my" way, then use the builder back to get the object. This way I'm sure everything will be replayable in the future, if I have coded "my" way correctly (with serials for examples).

This why I said wow!, this is great that this was already implemented in your code!
